src/storage/src/hummock/iterator/forward_user.rs

```diff
@@ -244,31 +198,87 @@ impl<I: HummockIterator<Direction = Forward>> UserIterator<I> {
         }
 
         if self.key_out_of_range() {
-            self.out_of_range = true;
             return Ok(());
         }
 
         self.delete_range_iter.seek(full_key.user_key).await?;
 
-        // Handle multi-version
-        self.last_key = FullKey::default();
-        // Handle range scan when key > end_key
-
-        self.next().await
+        self.try_advance_to_next_valid().await
     }
 
     /// Indicates whether the iterator can be used.
     pub fn is_valid(&self) -> bool {
-        // Handle range scan
-        // key >= begin_key is guaranteed by seek/rewind function
-        (!self.out_of_range) && self.iterator.is_valid()
+        self.is_current_pos_valid
     }
 
     pub fn collect_local_statistic(&self, stats: &mut StoreLocalStatistic) {
         stats.add(&self.stats);
         self.iterator.collect_local_statistic(stats);
     }
 
+    /// Advance the inner iterator to a valid position, in which the entry can be exposed.
+    /// Iterator will not be advanced if it already pointed to a valid position.
+    async fn try_advance_to_next_valid(&mut self) -> HummockResult<()> {
+        loop {
+            if !self.iterator.is_valid() {
+                break;
+            }
+
+            let full_key = self.iterator.key();
+            let epoch = full_key.epoch_with_gap.pure_epoch();
+
+            // Handle epoch visibility
+            if epoch < self.min_epoch || epoch > self.read_epoch {
+                self.iterator.next().await?;
+                continue;
+            }
+
+            // It is better to early return here if the user key is already
+            // out of range to avoid unnecessary access on the range tomestones
+            // via `delete_range_iter`.
+            // For example, if we are iterating with key range [0x0a, 0x0c) and the
+            // current key is 0xff, we will access range tombstones in [0x0c, 0xff],
+            // which is a waste of work.
+            if self.key_out_of_range() {
+                break;
+            }
+
+            // Skip older version entry for the same user key
+            if self.last_key.user_key.as_ref() == full_key.user_key {
```

Comment:
We can move this check to before `key_out_of_range` so that we can avoid always checking `key_out_of_range`, since `key_out_of_range` is much less likely to be hit than this equality check.

Reply:
If we limit that there is the only usage of `key_out_of_range`, in `key_out_of_range` we can assume that `self.last_key` is set with `full_key.copy_into()`, and then we can use the `self.last_key` to compare with the range instead of getting the key from iterator. Fetching the key from iterator may involve several layer of method call, and is slower than directly using the `self.last`